receiver/awscloudwatchreceiver/logs.go

```diff
@@ -0,0 +1,335 @@
+		}
+		l.namedPolls = namedPolls
+		l.prefixedPolls = prefixedPolls
+	}
+
+	for i := range l.namedPolls {
+		np := l.namedPolls[i]
+		l.wg.Add(1)
+		go l.poll(ctx, np.logGroup, &np)
+	}
+
+	for i := range l.prefixedPolls {
+		pp := l.prefixedPolls[i]
+		l.wg.Add(1)
+		go l.poll(ctx, pp.logGroup, &pp)
+	}
+
+	return nil
+}
+
+func (l *logsReceiver) Shutdown(ctx context.Context) error {
+	l.logger.Debug("shutting down logs receiver")
+	close(l.doneChan)
+	l.wg.Wait()
+	return nil
+}
+
+func (l *logsReceiver) poll(ctx context.Context, logGroup string, pc pollConfig) {
+	defer l.wg.Done()
+
+	t := time.NewTicker(l.pollInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-l.doneChan:
+			return
+		case <-t.C:
+			err := l.pollForLogs(ctx, logGroup, pc)
+			if err != nil {
+				l.logger.Error("there was an error during the poll", zap.String("log group", logGroup), zap.Error(err))
+			}
+		}
+	}
+}
+
+func (l *logsReceiver) pollForLogs(ctx context.Context, logGroup string, pc pollConfig) error {
+	err := l.ensureSession()
+	if err != nil {
+		return err
+	}
+	nextToken := aws.String("")
+	for nextToken != nil {
+		select {
+		// if done, we want to stop processing paginated stream of events
+		case _, ok := <-l.doneChan:
+			if !ok {
+				return nil
+			}
+		default:
+			startTime := time.Now().Add(-l.pollInterval)
+			endTime := time.Now()
+			input := pc.request(l.maxEventsPerRequest, &startTime, &endTime)
+			resp, err := l.client.FilterLogEventsWithContext(ctx, input)
+			if err != nil {
+				l.logger.Error("unable to retrieve logs from cloudwatch", zap.String("log group", logGroup), zap.Error(err))
+				break
+			}
+			observedTime := pcommon.NewTimestampFromTime(time.Now())
+			logs := l.processEvents(observedTime, logGroup, resp)
+			if logs.LogRecordCount() > 0 {
+				if err = l.consumer.ConsumeLogs(ctx, logs); err != nil {
+					l.logger.Error("unable to consume logs", zap.Error(err))
+					break
+				}
+			}
+			nextToken = resp.NextToken
+		}
+	}
+	return nil
+}
+
+func (l *logsReceiver) processEvents(now pcommon.Timestamp, logGroupName string, output *cloudwatchlogs.FilterLogEventsOutput) plog.Logs {
+	logs := plog.NewLogs()
+	for _, e := range output.Events {
+		if e.Timestamp == nil {
+			l.logger.Error("unable to determine timestamp of event as the timestamp is nil")
+			continue
+		}
+
+		if e.Message == nil {
+			l.logger.Error("no message was present on the event", zap.Any("event.id", e.EventId))
+			continue
+		}
+
+		rl := logs.ResourceLogs().AppendEmpty()
+		resourceAttributes := rl.Resource().Attributes()
+
+		safePutAttribute("aws.region", &l.region, resourceAttributes.PutString)
+		safePutAttribute("cloudwatch.log.group.name", &logGroupName, resourceAttributes.PutString)
+		safePutAttribute("cloudwatch.log.stream", e.LogStreamName, resourceAttributes.PutString)
+
+		logRecord := rl.ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
+		logRecord.SetObservedTimestamp(now)
```

Comment:
Double check for me, but I'm pretty sure this is auto-initialized. I don't think we need or necessarily want to align this value across the collection, as the reality is closer to just leaving it as the default.

Reply:
I just double checked and I don't think it does:

<details>

<summary> With SetObserveredTimestamp</summary>


```json
{
      "resource": {
        "attributes": [
          {
            "key": "aws.region",
            "value": {
              "stringValue": "us-east-2"
            }
          },
          {
            "key": "cloudwatch.log.group.name",
            "value": {
              "stringValue": "/aws/eks/dev-eks-0/cluster"
            }
          },
          {
            "key": "cloudwatch.log.stream",
            "value": {
              "stringValue": "kube-apiserver-audit-d89fc2f8d8934eda31d6dbc4e81e817d"
            }
          }
        ]
      },
      "scopeLogs": [
        {
          "scope": {},
          "logRecords": [
            {
              "timeUnixNano": "1665432688403000000",
              "observedTimeUnixNano": "1665432689314886000",
              "body": {
                "stringValue": "{\"kind\":\"Event\",\"apiVersion\":\"audit.k8s.io/v1\",\"level\":\"Metadata\",\"auditID\":\"9a4b39dd-de3d-45f9-a1d1-1dcc5140533f\",\"stage\":\"ResponseComplete\",\"requestURI\":\"/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/cloud-provider-extraction-migration?timeout=5s\",\"verb\":\"update\",\"user\":{\"username\":\"eks:cloud-controller-manager\",\"groups\":[\"system:authenticated\"]},\"sourceIPs\":[\"10.0.161.254\"],\"userAgent\":\"aws-cloud-controller-manager/v0.0.0 (linux/amd64) kubernetes/$Format/leader-election\",\"objectRef\":{\"resource\":\"leases\",\"namespace\":\"kube-system\",\"name\":\"cloud-provider-extraction-migration\",\"uid\":\"83a801b2-1631-4f13-a897-ac2977622315\",\"apiGroup\":\"coordination.k8s.io\",\"apiVersion\":\"v1\",\"resourceVersion\":\"28200671\"},\"responseStatus\":{\"metadata\":{},\"code\":200},\"requestReceivedTimestamp\":\"2022-10-10T20:11:28.299946Z\",\"stageTimestamp\":\"2022-10-10T20:11:28.303808Z\",\"annotations\":{\"authorization.k8s.io/decision\":\"allow\",\"authorization.k8s.io/reason\":\"RBAC: allowed by ClusterRoleBinding \\\"eks:cloud-controller-manager\\\" of ClusterRole \\\"eks:cloud-controller-manager\\\" to User \\\"eks:cloud-controller-manager\\\"\"}}"
              },
              "attributes": [
                {
                  "key": "id",
                  "value": {
                    "stringValue": "37140390029379149665944061435001139263417947014350635033"
                  }
                }
              ],
              "traceId": "",
              "spanId": ""
            }
          ]
        }
      ]
    }

```
</details>

<details>
<summary> Without `setObservedTimestamp` </summary>

```json
    {
      "resource": {
        "attributes": [
          {
            "key": "aws.region",
            "value": {
              "stringValue": "us-east-2"
            }
          },
          {
            "key": "cloudwatch.log.group.name",
            "value": {
              "stringValue": "/aws/eks/dev-eks-0/cluster"
            }
          },
          {
            "key": "cloudwatch.log.stream",
            "value": {
              "stringValue": "kube-apiserver-audit-d89fc2f8d8934eda31d6dbc4e81e817d"
            }
          }
        ]
      },
      "scopeLogs": [
        {
          "scope": {},
          "logRecords": [
            {
              "timeUnixNano": "1665432815710000000",
              "body": {
                "stringValue": "{\"kind\":\"Event\",\"apiVersion\":\"audit.k8s.io/v1\",\"level\":\"Metadata\",\"auditID\":\"378931e9-b75e-4eb5-98c2-b76e785db23f\",\"stage\":\"ResponseComplete\",\"requestURI\":\"/api/v1/namespaces/kube-system/configmaps/eks-certificates-controller\",\"verb\":\"update\",\"user\":{\"username\":\"eks:certificate-controller\",\"groups\":[\"system:authenticated\"]},\"sourceIPs\":[\"10.0.161.254\"],\"userAgent\":\"eks-certificates-controller/v0.0.0 (linux/amd64) kubernetes/$Format\",\"objectRef\":{\"resource\":\"configmaps\",\"namespace\":\"kube-system\",\"name\":\"eks-certificates-controller\",\"uid\":\"a39c0df1-3c0c-417b-bd98-9ef28131f8d1\",\"apiVersion\":\"v1\",\"resourceVersion\":\"28201015\"},\"responseStatus\":{\"metadata\":{},\"code\":200},\"requestReceivedTimestamp\":\"2022-10-10T20:13:35.582034Z\",\"stageTimestamp\":\"2022-10-10T20:13:35.587283Z\",\"annotations\":{\"authorization.k8s.io/decision\":\"allow\",\"authorization.k8s.io/reason\":\"RBAC: allowed by RoleBinding \\\"eks:certificate-controller/kube-system\\\" of Role \\\"eks:certificate-controller\\\" to User \\\"eks:certificate-controller\\\"\"}}"
              },
              "attributes": [
                {
                  "key": "id",
                  "value": {
                    "stringValue": "37140392868420118655282101869312050893585421491989839889"
                  }
                }
              ],
              "traceId": "",
              "spanId": ""
            }
          ]
        }
      ]
    }
```

</details>

Note the missing "observedTimeUnixNano" parameter in the without clause.

Reply:
Ok, thanks for double checking. I was thinking of how is _should_ work according to the spec, but you were right in the first place that it is up to the implementation to make it work that way.